src/mon/OSDMonitor.cc

```diff
@@ -2640,6 +2640,25 @@ bool OSDMonitor::preprocess_query(MonOpRequestRef op)
   Message *m = op->get_req();
   dout(10) << "preprocess_query " << *m << " from " << m->get_orig_source_inst() << dendl;
 
+  // Clean up the records, because we have received its message
+  auto s = m->get_orig_source();
+  if (s.is_osd()) {
+    auto sid = s.num();
+    if (failure_info.count(sid)) {
+      failure_info_t& fi = failure_info[sid];
+      for (auto& [reporter, fr] : fi.reporters) {
+	MonOpRequestRef report_op = fi.cancel_report(reporter);
+	if (report_op) {
+	  mon->no_reply(report_op);
+	}
+      }
+      ceph_assert(fi.reporters.empty());
+      dout(10) << " removing last failure_info for osd." << sid
+	       << dendl;
+      failure_info.erase(sid);
+    }
+  }
+
```

Comment:
i failed to find any analysis or reasonings in the commit message or in the tracker ticket. why do you think the peer osds cannot withdraw the failure report from monitor after the failed one is reachable again?

Reply:
@tchaikov  I find the issue and a path to fix it.
I think this problem still exists. 
You can see this patch for some information.
https://github.com/ceph/ceph/pull/28177

I think we need to clean up the records of the osd if  we have received its message.

Reply:
@simon-rock i tried to contact you via your mail address in the Signed-off-by line, but no replies so far. my question still holds.

Reply:
i think we should not rely on the message from the target osd to cancel the stale reports. what if the target osd is relatively quiet from the leader's perspective after being reported unresponsive?

i am creating #40033 as an alternative to address this issue. to check the `failed_for` time, if it takes way to long to mark an osd down, we just give up and drop all failure reports.

Reply:
discussed with @simon-rock offline. this approach is better. but couple nits

- better off cancel the failure report in `preprocess_beacon()`. as an active osd is supposed to send beacon messages periodically.
- no need to reply no_op
- no need to lookup in `failure_info` twice.